Moving docker service digest pinning to client side

nishanttotla opened the pull request:
This PR is an alternative proposal to #32384, based on discussions with @aluzzardi @dhiltgen @alexmavr. 

The _heavy lifting_ in this case happens in the `client` package, which is easier to use for third-party clients, and they get the benefit of client-side digest pinning without having to do it themselves.

cc @aaronlehmann 

**- What I did**
Digest pinning was introduced for Docker services in #28173, and it was performed on the daemon. After #32061, I moved digest pinning to the client, simplifying it.

**- How I did it**
The client makes a call to `/images/image_name/manifest` to retrieve manifest information, which contains the image digest.

**- How to verify it**
`docker service create --name test alpine sleep 1000` then `docker service inspect test` should contain an image with the digest.

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->
Docker service image pin by digest moved to client side.

**- WIP**
- [x] Merge #32061
- [x] Return warning when pinning fails
- [x] Disable daemon side digest pinning for new API versions (unclear if this is necessary)

**- A picture of a cute animal (not mandatory but encouraged)**
![goshi](https://cloud.githubusercontent.com/assets/1872537/24725003/22d41800-1a02-11e7-8d73-b81a4af06214.jpg)

This is Gosha, the friendliest cat in Berkeley.

nishanttotla committed:
Moving docker service digest pinning to client side

Signed-off-by: Nishant Totla <nishanttotla@gmail.com>

alexmavr commented on client/service_create.go:
```diff
@@ -18,13 +20,28 @@ func (cli *Client) ServiceCreate(ctx context.Context, service swarm.ServiceSpec,
 
+	// Call the /manifest endpoint to retrieve manifest information
+	var manifestInspect registrytypes.ManifestInspect
+	manifestResp, manifestErr := cli.get(ctx, "/images/"+service.TaskTemplate.ContainerSpec.Image+"/manifest", url.Values{}, headers)
```
What if the `Image` is already a manifest?

nishanttotla replied:
You mean the image already contains a digest? Then the digest will be used to retrieve the manifest, and platform information added (this is not part of this PR, and will come later in a separate PR). Without the platform information stuff, we could just skip this call altogether.

aaronlehmann commented:
I'm okay with this approach but I think we should add a flag in `ServiceCreateOptions` to let API users opt in or opt out. Otherwise it would be impossible to use the client to create a service that isn't pinned to a digest.

nishanttotla commented:
@aaronlehmann I agree, I'll add that. Are there any cons to this approach, apart from having to move digest pinning away from content-trust?

aaronlehmann commented:
I think it's problematic for the "service update" case. We should only be messing with the image field or the platform info if the user actually specified `--image` on the command line - it shouldn't be a side effect of an unrelated change like scaling the service. In the CLI it's easy to gate this stuff with an `if flags.Changed("image")`, but if it's in the client and it's something we have to explicitly enable or disable, it seems a bit more awkward. Not a dealbreaker but I think it's definitely awkward for the update case if we do it this way.

What if we moved `resolveServiceImageDigest` from the other PR into `client`?

nishanttotla commented:
@aaronlehmann I'll update `ServiceUpdateOptions` to include a field indicating `if flags.Changed("image")`.

When you say moving `resolveServiceImageDigest` to the `client` package, do you also want to move content trust here? It doesn't look difficult to do, and also alleviates some of my concerns with this PR separating digest pinning and content trust into separate places.

aaronlehmann commented:
> When you say moving `resolveServiceImageDigest` to the `client` package, do you also want to move content trust here? It doesn't look difficult to do, and also alleviates some of my concerns with this PR separating digest pinning and content trust into separate places.

I'm not sure. It sounds like a good idea if it's possible. But I seem to remember that content trust involves a lot of code that lives in `cli`.

nishanttotla commented:
@aaronlehmann let me take a look. If it's easy to refactor, I'll include it in this PR.

nishanttotla commented:
Update: Trust code is definitely entrenched in the CLI, and while it seems "refactorable", I don't think it should be a part of this PR because it is associated with content trust for running containers as well.

nishanttotla commented:
@aaronlehmann: updated this PR. Added a field called `UpdateImage` to `ServiceCreateOptions` that gets set when the `--image` flag is provided. In the future, it can also be easily used to disable digest pinning from the client.

I also updated `resolveServiceImageDigest` to `resolveServiceImageDigestContentTrust` to indicate that it only resolves digest by content trust. In the future, we should move this function to the `client` package.

aaronlehmann commented on api/types/client.go:
```diff
@@ -325,6 +325,12 @@ type ServiceUpdateOptions struct {
+	// the image. This is used to decide if the image needs to be pinned
+	// by digest. The field is set when the `--image` flag is provided,
+	// or the user requests explicitly to disable pin by digest.
+	UpdateImage bool
```
Let's name this something like `QueryRegistry`, since I guess it will be extended to things other than pinning by digest in the future (platform info...).

I think we should support the same flag for `ServiceCreate`, so that digest pinning is not mandatory for all `client` users.

Let's not mention the `--image` flag (except possibly as an example), because the client is independent of the Docker CLI.

aaronlehmann commented on client/service_create.go:
```diff
@@ -18,13 +18,25 @@ func (cli *Client) ServiceCreate(ctx context.Context, service swarm.ServiceSpec,
+	manifestInspect, manifestErr := cli.ImageManifest(ctx, service.TaskTemplate.ContainerSpec.Image, options.EncodedRegistryAuth)
+	if manifestErr == nil {
+		// manifest was successfully retrieved, now pin by digest
+		service.TaskTemplate.ContainerSpec.Image = string(manifestInspect.Digest)
```
We must not overwrite the `Image` field here if it already contains a digest (for example, if digest pinning was already done through content trust).

aaronlehmann commented on client/service_update.go:
```diff
@@ -33,13 +33,29 @@ func (cli *Client) ServiceUpdate(ctx context.Context, serviceID string, version
+		manifestInspect, manifestErr := cli.ImageManifest(ctx, service.TaskTemplate.ContainerSpec.Image, options.EncodedRegistryAuth)
+		if manifestErr == nil {
+			// manifest was successfully retrieved, now pin by digest
+			service.TaskTemplate.ContainerSpec.Image = string(manifestInspect.Digest)
```
We must not overwrite the `Image` field here if it already contains a digest (for example, if digest pinning was already done through content trust).

nishanttotla commented:
@aaronlehmann: addressed your comments.

aaronlehmann commented on client/service_create.go:
```diff
@@ -3,28 +3,48 @@ package client
+		if manifestErr == nil {
+			// manifest was successfully retrieved, now pin by digest if the
+			// image doesn't already contain a digest
+			if _, err := reference.ParseNamed(service.TaskTemplate.ContainerSpec.Image); err != nil {
```
This doesn't seem right. We should be replacing the `Image` field if the image reference doesn't have a digest, not if it doesn't parse.

nishanttotla replied:
The reasoning for this was to not update the image if it parses to a legitimate canonical reference AND a manifest is retrieved. Is that not what this is doing? I wanted to reuse something from the `distribution/reference` package. Do you recommend something else?

aaronlehmann replied:
Where does it check that the reference is canonical?

nishanttotla replied:
```
func ParseNamed(s string) (Named, error) {
	named, err := ParseNormalizedNamed(s)
	if err != nil {
		return nil, err
	}
	if named.String() != s {
		return nil, ErrNameNotCanonical
	}
	return named, nil
}
```
This check I believed was doing that, but perhaps I'm getting it wrong.
```
f named.String() != s {
		return nil, ErrNameNotCanonical
	}
```
Adding a type check for `reference.Canonical` should fix that?

aaronlehmann replied:
That "canonical" is something else :(

> Adding a type check for `reference.Canonical` should fix that?

Yes

nishanttotla replied:
Ah okay, let me add that. I wasn't going by the word "canonical" as much as the string check, but it's embarrassing now because that doesn't make any sense. Fixing it now.

nishanttotla replied:
Fixed.

clnperez commented:
Since @dnephin mentioned that this my conflict with my PR (docker/docker/pull/27455), @nishanttotla can you clarify whether the imageManifest func you're adding pulls from the registry or expects the image to be pre-pulled?

nishanttotla commented:
@clnperez it talks to the registry directly, does not look at local images at all.

nishanttotla commented:
Resolved the merge conflict, and updated the PR to reflect changes in https://github.com/moby/moby/pull/32061

nishanttotla commented:
Moving CLI changes to https://github.com/docker/cli/pull/30 and rebasing this PR.

nishanttotla commented:
Rebased on top of #32061.

Ping @aaronlehmann @aluzzardi

aaronlehmann commented:
Is it still WIP?

nishanttotla commented:
@aaronlehmann The only WIP part is to return the warning, for which I need to update a couple of lines. I'm working on making sure this works as expected end to end right now.

The other thing was to decide whether we should disable digest pinning in the daemon. There are arguments both ways for that.

nishanttotla commented:
This is no longer WIP. The relevant CLI changes are in https://github.com/docker/cli/pull/30, including a temporary vendor commit to help CI pass.

Just one decision to make regarding the digest pinning code in the daemon though.
- We disable digest pinning in the daemon for older API versions, that way everything works fine for API version 1.30 and onwards. I prefer this.
- We don't disable digest pinning in the daemon. Right now, a double warning is returned when pinning fails (one from the daemon, one from the client). We ignore all warnings returned from the daemon and overwrite them with the client warnings. This is problematic because some other feature might use the `Warnings` array in the future, and we don't want to overwrite them. Also there's the extra cost of making a second trip to the registry for no reason.

So my question is, is there any reason to not disable digest pinning for older API versions? The only argument I've heard is that people usually don't provide API versions at all (so that defaults to latest), and continue updating their daemon. This way they don't get digest pinning at all. But of course not using specific API versions is path to issues in production and broken dreams.

@aaronlehmann @aluzzardi @alexmavr WDYT?

aaronlehmann commented:
Let's disable the code path in the daemon for older API versions.

nishanttotla committed:
Disabling digest pinning for API versions < 1.30

Signed-off-by: Nishant Totla <nishanttotla@gmail.com>

nishanttotla commented:
Disabled digest pinning for client versions < 1.30.

aaronlehmann commented on client/service_create.go:
```diff
@@ -2,29 +2,84 @@ package client
+		}
+	} else {
+		headers = map[string][]string{
+			"version": {cli.version},
```
```go
headers := map[string][]string{
	"version": {cli.version},
}

if options.EncodedRegistryAuth != "" {
	headers["X-Registry-Auth"] = {options.EncodedRegistryAuth}
}
```

aaronlehmann commented on client/service_update.go:
```diff
@@ -15,13 +15,21 @@ func (cli *Client) ServiceUpdate(ctx context.Context, serviceID string, version
 		}
 	}
+	// update CLI version
+	headers["version"] = []string{cli.version}
```
`headers["version"]` was already set above. Suggest initializing the map with `version` already set, and simplifying the conditionals above.

nishanttotla replied:
AH, I meant to delete this line, sorry.

aaronlehmann commented on client/service_create.go:
```diff
@@ -2,29 +2,80 @@ package client
-	var headers map[string][]string
+	var (
+		distErr error
+	)
```
Can be `var distErr error` (but it doesn't matter).

aaronlehmann commented on daemon/cluster/services.go:
```diff
@@ -151,8 +151,8 @@ func (c *Cluster) CreateService(s types.ServiceSpec, encodedAuth string) (*apity
-			// pin image by digest
-			if os.Getenv("DOCKER_SERVICE_PREFER_OFFLINE_IMAGE") != "1" {
+			// pin image by digest for API versions < 1.30
+			if os.Getenv("DOCKER_SERVICE_PREFER_OFFLINE_IMAGE") != "1" && queryRegistry {
```
Is there a good reason to continue supporting `DOCKER_SERVICE_PREFER_OFFLINE_IMAGE`? This was for the integration tests, but if the integration tests use the latest API version, they shouldn't hit this code path anyway.

That said, it may take some work to get things right with the integration tests after removing this. It may be best to remove it in a followup. Speaking of which, how will we prevent CLI-oriented integration tests from trying to resolve digests for local images on Docker Hub? Should we add a CLI option to prevent digest pinning?

nishanttotla replied:
This seems correct, we shouldn't have to keep `DOCKER_SERVICE_PREFER_OFFLINE_IMAGE`, but it does absolutely no harm right now, and isn't user facing, so we should just wait on removing it. I'll add a comment indicating this.

If the CLI-oriented tests will purely run CLI commands, then we will have to add a flag, there is no other way. Given that we know of scenarios where disabling is required, we should add the flag.

aaronlehmann commented:
Mostly LGTM, but a little concerned that moving digest pinning to the client side will break the `DOCKER_SERVICE_PREFER_OFFLINE_IMAGE` workaround, and cause integration tests that use `docker service create` or `docker service update --image` in the CLI to interact with the hub when they really should not. I think we either need to add a flag to disable digest pinning in the CLI, or rewrite the applicable integration tests to avoid the CLI.

nishanttotla commented:
@aluzzardi what do you think about adding a flag? It shouldn't be hard to do, given that this PR already has an API-level option to disable the registry lookup.

aluzzardi commented:
@nishanttotla @aaronlehmann A flag sounds great, I think it was one of the goals of moving resolution client side.

May I suggest `--no-resolve-image` or something like that?

nishanttotla commented:
@aluzzardi @aaronlehmann I'll open a follow-up PR for that, since most changes will be in `docker/cli`, and we might need some time to agree on the flag name and behavior.

aaronlehmann commented:
That sounds okay. Hopefully the circular dependency between `moby` and `cli` doesn't put us in a state where integration tests are trying to perform digest pinning (which makes them flaky). Ideally the commit that updates `moby` to point to a revision of `cli` that does the digest pinning will also update the integration tests to use the new flag that disables it.

nishanttotla commented:
I'll add the flag to the same PR (https://github.com/docker/cli/pull/30) that enables the `QueryRegistry` option in the CLI.

aaronlehmann commented:
LGTM

tiborvass commented on client/service_create.go:
```diff
@@ -2,29 +2,78 @@ package client
+// image name that could not be pinned by digest. The formatting
+// is hardcoded, but could me made smarter in the future
+func digestWarning(image string) string {
+	return fmt.Sprintf("image %s could not be accessed on a registry to record\nits digest. Each node will access %s independently,\npossibly leading to different nodes running different\nversions of the image.\n", image, image)
```
the `\n` in the middle like that is kinda unusual

nishanttotla replied:
@tiborvass this is to add line breaks, since the text is big. This is how it is also in the daemon right now.

tiborvass commented on api/types/client.go:
```diff
@@ -276,6 +276,12 @@ type ServiceCreateOptions struct {
+	// contacting a registry. A registry may be contacted to retrieve
+	// the image digest and manifest, which in turn can be used to update
+	// platform or other information about the service.
+	QueryRegistry bool
```
@nishanttotla I'm wondering if the default shouldn't be the current behavior. `SkipRegistryQuery` ?

nishanttotla replied:
@tiborvass the corresponding CLI PR is https://github.com/docker/cli/pull/30.

The idea is that `QueryRegistry` should be set on all `service create` commands unless the user explicitly provides a flag to prevent it, and for all `service update` commands that update the image, using the `--image` flag (also unless the user explicitly asks to not do it with the flag).

aaronlehmann replied:
@tiborvass: The current behavior is that the client does not perform this query.

tiborvass replied:
@aaronlehmann my bad i thought this was server side.

tiborvass commented:
LGTM

nishanttotla commented:
Adding link back to https://github.com/moby/moby/issues/31348
